services/continuousquery/service.go

```diff
@@ -0,0 +1,309 @@
+/*
+Copyright 2022 Huawei Cloud Computing Technologies Co., Ltd.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+ http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package continuousquery
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+	
+	log "github.com/influxdata/influxdb/logger"
+	query2 "github.com/influxdata/influxdb/query"
+	"github.com/openGemini/openGemini/engine/index/tsi"
+	"github.com/openGemini/openGemini/lib/statisticsPusher/statistics"
+	"github.com/openGemini/openGemini/open_src/influx/influxql"
+	"github.com/openGemini/openGemini/open_src/influx/meta"
+	"github.com/openGemini/openGemini/open_src/influx/query"
+	"github.com/openGemini/openGemini/services"
+	"go.uber.org/zap"
+)
+
+const (
+	// DefaultReportTime is the default time interval for reporting stats.
+	DefaultReportTime     = 5 * time.Minute
+	DefaultInnerChunkSize = 1024
+)
+
+// ContinuousQuery is a struct for cq info, LastRun etc.
+type ContinuousQuery struct {
+	Database string
+	Info     *meta.ContinuousQueryInfo
+	HasRun   bool
+	LastRun  time.Time
+	q        *influxql.SelectStatement
+}
+
+// Service include query and store engine for CQ
+type Service struct {
+	services.Base
+	
+	MetaClient interface {
+		Databases() map[string]*meta.DatabaseInfo
+	}
+	
+	Engine interface {
+	}
+	
+	mu            sync.RWMutex
+	lastRuns      map[string]time.Time
+	stop          chan struct{}
+	wg            *sync.WaitGroup
+	QueryExecutor *query.Executor
+	RunInterval   time.Duration
+}
+
+// NewService creates a new Service instance named continuousQuery
+func NewService(interval time.Duration) *Service {
+	s := &Service{
+		lastRuns:    map[string]time.Time{},
+		RunInterval: DefaultReportTime,
+	}
+	s.Init("continuousQuery", interval, s.handle)
+	return s
+}
+
+func (s *Service) handle() {
+	_, logEnd := log.NewOperation(s.Logger.GetZapLogger(), "continuous query check", "continuousQuery_check")
+	defer logEnd()
+	s.Logger.Info("Starting continuous query service")
+	
+	s.stop = make(chan struct{})
+	s.wg = &sync.WaitGroup{}
+	s.wg.Add(1)
+	go s.backgroundLoop()
```

Comment:
`s.backgroundLoop()` is enough. Otherwise, the `handle()` will be repeatedly called.